pym/portage/meta-manifest.py

```diff
@@ -0,0 +1,276 @@
+        'portage.util:write_atomic,writemsg_level',
+)
+
+from gkeys.gkeysinterface import GkeysInterface
+from gkeys.config import GKeysConfig
+from pyGPG.gpg import GPG
+from gkeys.lib import GkeysGPG
+
+from portage import manifest
+from portage import os
+from portage import _encodings
+from portage import _unicode_decode
+from portage import _unicode_encode
+from portage.manifest import guessManifestFileType
+from portage.exception import DigestException, FileNotFound, \
+        InvalidDataType, MissingParameter, PermissionDenied, \
+        PortageException, PortagePackageException
+from portage.const import (MANIFEST1_HASH_FUNCTIONS, MANIFEST2_HASH_DEFAULTS,
+        MANIFEST2_HASH_FUNCTIONS, MANIFEST2_IDENTIFIERS, MANIFEST2_REQUIRED_HASH)
+from portage.localization import _
+
+if sys.hexversion >= 0x3000000:
+	# pylint: disable=W0622
+	_unicode = str
+	basestring = str
+else:
+	_unicode = unicode
+
+
+class MetaManifest(manifest.Manifest):
+	'''Subclass of the Manifest class for Manifest files outside the packages'''
+
+	def sign(self):
+		'''Signs MetaManifest file with the default PORTAGE_GPG_KEY''' 
+		filename = self.getFullname()
+		portage_settings = portage.config(clone=portage.settings)
+		portage_portdbapi = portage.portdbapi(portage_settings)
+		gpgcmd = portage_settings.get("PORTAGE_GPG_SIGNING_COMMAND")
+		if gpgcmd in [None, '']:
+			raise portage.exception.MissingParameter(
+				"PORTAGE_GPG_SIGNING_COMMAND is unset! Is make.globals missing?")
+		if "${PORTAGE_GPG_KEY}" in gpgcmd and "PORTAGE_GPG_KEY" not in portage_settings:
+			raise portage.exception.MissingParameter("PORTAGE_GPG_KEY is unset!")
+		if "${PORTAGE_GPG_DIR}" in gpgcmd:
+			if "PORTAGE_GPG_DIR" not in portage_settings:
+				portage_settings["PORTAGE_GPG_DIR"] = os.path.expanduser("~/.gnupg")
+			else:
+				portage_settings["PORTAGE_GPG_DIR"] = os.path.expanduser(portage_settings["PORTAGE_GPG_DIR"])
+			if not os.access(portage_settings["PORTAGE_GPG_DIR"], os.X_OK):
+				raise portage.exception.InvalidLocation(
+					"Unable to access directory: PORTAGE_GPG_DIR='%s'" %
+					portage_settings["PORTAGE_GPG_DIR"])
+		gpgvars = {"FILE": filename}
+		for setting in ("PORTAGE_GPG_DIR", "PORTAGE_GPG_KEY"):
+			keyvar = portage_settings.get(setting)
+			if keyvar is not None:
+				gpgvars[setting] = keyvar
+		gpgcmd = portage.util.varexpand(gpgcmd, mydict=gpgvars)
+		gpgcmd = portage.util.shlex_split(gpgcmd)
+		gpgcmd = [portage._unicode_encode(arg, encoding=portage._encodings['fs'], errors='strict') for arg in gpgcmd]
+		return_code = subprocess.call(gpgcmd)
+		if return_code == os.EX_OK:
+			os.rename(filename + ".asc", filename)
+		else:
+			raise portage.exception.PortageException("!!! gpg exited with '" + str(return_code) + "' status")
+
+	def find_reporoot(self):
+		'''Finds reporoot'''
+		repodir = self.pkgdir
+		while len(set(['metadata', 'eclass', 'profiles']).intersection(os.listdir(repodir))) == 0:
```

Comment:
Since os.listdir is more expensive than stat, and intersection is more expensive than any(), this would be better:
```
while not any(os.path.exists(os.path.join(repodir, name)) for name in ('metadata', 'eclass', 'profiles'))):
```